jason/src/utils/component.rs

```diff
@@ -0,0 +1,101 @@
+//! Implementation of the [`Component`].
+
+use std::{cell::RefCell, ops::Deref, rc::Rc};
+
+use futures::{future, Future, FutureExt, Stream, StreamExt};
+use wasm_bindgen_futures::spawn_local;
+
+use crate::utils::{JasonError, TaskHandle};
+
+/// Creates and spawns new [`Component`].
+///
+/// `$component` - type alias for [`Component`] which you wanna create.
+///
+/// `$state` - [`Component`]'s state.
+///
+/// `$obj` - object managed by [`Component`].
+#[macro_export]
+macro_rules! spawn_component {
+    ($component:ty, $state:expr, $obj:expr $(,)*) => {{
+        let component = <$component>::inner_new($state, $obj);
+        component.spawn();
+        component
```

Comment:
@evdokimovs @alexlapa I didn't get it.

If the lifecycle of `spawn` is strictly bound to the `Component` lifetime, why bother with hidden and macro? Isn't it better just to have visible constructor and do the spawn inside it? Without the need to mess around?

Reply:
Oh, I see... it's because `spawn()` method is generated by a macro for concrete types.

Ok, but why bother with a macro at all here? Isn't a regular trait is sufficient here?

```rust
trait WatcherComponent<S: Stream> {
    fn watch(&self) -> S;

    async fn change(
        ctx: Rc<Sender>,
        state: Rc<SenderState>,
         new_val: S::Item
     ) -> Result<(), ()>;
}
```

Reply:
@tyranron ,

> Oh, I see... it's because `spawn()` method is generated by a macro for concrete types.
> 
> Ok, but why bother with a macro at all here? Isn't a regular trait is sufficient here?
> 
> ```rust
> trait WatcherComponent<S: Stream> {
>     fn watch(&self) -> S;
> 
>     async fn change(
>         ctx: Rc<Sender>,
>         state: Rc<SenderState>,
>          new_val: S::Item
>      ) -> Result<(), ()>;
> }
> ```

Я уже пытался избавиться от `spawn_component!`, но ни один из придуманных вариантов с трейтами не подходил и был сильно сложнее и запутанее, чем простой макрос.

В Вашем варианте не ясно, что делать в случае если есть два обработчика для одного и того же типа. Например:

```rust
struct State {
    foo: ObservableCell<bool>,
    bar: ObservableCell<bool>,
}

#[watchers]
impl Component {
    #[watch(self.state().foo.subscribe()]
    fn foo_watcher(ctx: Rc<Ctx>, state: Rc<State>, val: bool) -> Result<(), ComponentError> {
        // ...
    }

    #[watch(self.state().bar.subscribe()]
    fn bar_watcher(ctx: Rc<Ctx>, state: Rc<State>, val: bool) -> Result<(), ComponentError> {
        // ...
    }
}
```

Reply:
@evdokimovs ok, now it's different. From the examples in code it wasn't obvious that `spawn()` method can have multiple `spawn_watcher()` calls. I have assumed that they come in `spawn()` + `watcher()` _pairs only_.

That sort of thing cannot be expressed at type-level at all, until we have some sort of variadic generics. Another way can be try HList, but that is an advanced thing and will only complicate the whole thing.

So, yes. Given this requirement, the macro is the only reasonable enough way, and we don't have any other options.

Please, update the examples in code with 2 `#[watch]` methods, so this will be a visible and explicit requirement.

Reply:
@tyranron ,

Я вроде-бы, нашел вариант избавиться от `spawn_component!` макроса, и сделать так, чтобы было невозможно создать компонент не заспавнив его watchers. Но это создаст ограничение из-за которого watcher'ы смогут подписываться только на поля из стейта. На мой взгляд ничего плохого в этом ограничении нет, но что будет в будущем никто не знает.

Если вкратце, то добавляется вот такой трейт:

```rust
/// Abstraction for the all states of the [`Component`].
pub trait ComponentState<C>: Sized {
    /// Spawns all watchers required for this [`ComponentState`].
    fn spawn_watchers(&self, watchers_spawner: &mut WatchersSpawner<Self, C>);
}
```

Этот трейт должен будет имплементиться для всех стейтов. Функция `Component::inner_new` выпиливается вместе с макросом `spawn_component!`, вместов них добавляется функция `Component::new`:

```rust
impl<S: ComponentState<C> + 'static, C: 'static> Component<S, C> {
    /// Returns new [`Component`] with a provided context and state.
    ///
    /// Spawns all watchers of this [`Component`].
    pub fn new(ctx: Rc<C>, state: Rc<S>) -> Self {
        let mut watchers_spawner =
            WatchersSpawner::new(Rc::clone(&state), Rc::clone(&ctx));
        state.spawn_watchers(&mut watchers_spawner);

        Self {
            state,
            ctx,
            _spawned_watchers: watchers_spawner.finish(),
        }
    }
}
```

Таким образом единственным вариантом создать компонент становится вызов `Component::new`, который и будет спавнить всех watcher'ов.

сс @alexlapa

Reply:
@evdokimovs we just move the responsibility from components to states. If we could derive the `ComponentState` implementation easily, then it will be affordable, otherwise we hit ergonomics largerly.

Reply:
@tyranron ,

Предполагается, что имплементация `ComponentState` будет генерится с помощью макроса `#[watchers]`, соответственно `ComponentState` в ручную трогать никто не будет. Но тогда в макросе придется еще и указывать тип контекста.

Вот пример того как тогда может выглядеть `#[watchers]`:

```rust
#[watchers(Ctx)]
impl State {
    #[watch(self.foo.subscribe()]
    fn foo_watcher(ctx: Rc<Ctx>, state: Rc<State>, val: bool) -> Result<(), ComponentError> {
        // ...
    }

    #[watch(self.bar.subscribe()]
    fn bar_watcher(ctx: Rc<Ctx>, state: Rc<State>, val: bool) -> Result<(), ComponentError> {
        // ...
    }
}
```

И генерить он будет имплементацию `ComponentState`:

```rust
impl ComponentState<Ctx> for State {
    fn spawn_watchers(&self, s: &mut WatchersSpawner<Self, Ctx>) {
        s.spawn(
            self.foo.subscribe(),
            Self::foo_watcher,
        );
        s.spawn(
            self.bar.subscribe(),
            Self::bar_watcher,
        );
    }
}
```

Самое неприятное тут это то, что теперь watcher'ы вешаются на `State`, и выглядеть все стало нелогичней. Поэтому, если `spawn_component!` не смущает, то и смысла это делать нет.